### PR TITLE
Turn autocast off when precision is fp32

### PR DESCRIPTION
**Pull request by aklife97**
# What does this PR do ?

Fix for autocast when precision is fp32

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 1]
```diff
@@ -311,7 +311,7 @@ def training_step(self, dataloader_iter, batch_idx):
             dtype=self.autocast_dtype,
             grad_scaler=self.trainer.precision_plugin.scaler.scale if self.cfg.precision == 16 else None,
             sequence_parallel=self.cfg.get('sequence_parallel', False),
-            enable_autocast=True,
+            enable_autocast=False if self.autocast_dtype == torch.float else True,
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:314]: This should be False if we're using O2 as well. Maybe just make a flag in the constructor that sets this true only when not using O2 and and when precision is 16 or bf16.

Comment (aklife97) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:314]: okay, but we might have to test that it doesn't break some random niche thing
Previously with Apex, autocast was being set even with O2

Comment (aklife97) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py]: done!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**